chore(transcription): change transcribers transcribe method signature

lutangar opened the pull request:
Introduce whisper builtin model.

## Description

<!-- Please include a summary of the change, with motivation and context -->

## Related issues

<!-- If suggesting a new feature or change, please discuss it in an issue first -->
<!-- If fixing a bug, there should be an issue describing it with steps to reproduce -->

## Has this been tested?

<!-- Put an `x` in the box that applies: -->
<!-- Check the unit test guide: https://docs.joinpeertube.org/contribute-getting-started?id=unit-tests -->

- [ ] 👍 yes, I added tests to the test suite
- [ ] 💭 no, because this PR is a draft and still needs work
- [ ] 🙅 no, because this PR does not update server code
- [ ] 🙋 no, because I need help <!-- Detail how we can help you -->

## Screenshots

<!-- delete if not relevant -->

lutangar committed:
chore(transcription): change transcribers transcribe method signature

Introduce whisper builtin model.

lutangar committed:
fix(transcription): activate language detection

Forbid transcript creation without a language.
Add `languageDetection` flag to an engine and some assertions.

Fix an issue in `whisper-ctranslate2` :
https://github.com/Softcatala/whisper-ctranslate2/pull/93

lutangar committed:
chore(transcription): use PeerTube time helpers instead of custom ones

Update existing time function to output an integer number of seconds and add a ms human-readable time formatter with hints of tests.

lutangar committed:
chore(transcription): use PeerTube UUID helpers

lutangar committed:
chore(transcription): enable CER evaluation

Thanks to this recent fix in Jiwer <3
https://github.com/jitsi/jiwer/issues/873

lutangar committed:
chore(transcription): creates JiWer package

I'm not very happy with the TranscriptFileEvaluator constructor... suggestions ?

chore(JiWer): add usage in README

docs(jiwer): update JiWer readme

lutangar committed:
chore(transcription): use FunMOOC video in fixtures

lutangar committed:
chore(transcription): add proper english video fixture

lutangar committed:
chore(transcription): use os tmp directory where relevant

lutangar committed:
chore(transcription): move benchmark out of tests

chore(transcription): remove transcription workflow

docs(transcription): add benchmark info

Chocobozzz commented on .github/workflows/transcription.yml:
```diff
@@ -29,6 +32,7 @@ jobs:
       - name: Run transcription tests
         run: |
           npm run mocha -- --exit --bail "packages/tests/src/transcription/**/*.spec.ts"
+          yarn run benchmark-transcription
```
I think we can remove the benchmark from the CI

lutangar replied:
I might as well remove the `.github/workflows/transcription.yml` file altogether.  What do you think ?

lutangar replied:
As for the Python dependencies I'll add `"preinstall": "pip install -r requirements.txt"` to the transcription package.

Chocobozzz replied:
> I might as well remove the .github/workflows/transcription.yml file altogether. What do you think ?

Yes please!

lutangar replied:
Not sure about the "test" workflow :
- which test suite will run the transcription package test ?
- shall I execute the preinstall script manually somewhere ?

Chocobozzz replied:
You can create another test case here: https://github.com/Chocobozzz/PeerTube/blob/develop/scripts/ci.sh where you install your dependencies 

Then just add the it to the matrix: https://github.com/Chocobozzz/PeerTube/blob/develop/.github/workflows/test.yml#L42

Chocobozzz commented on packages/jiwer/README.md:
```diff
@@ -0,0 +1,9 @@
+> *JiWER is a python tool for computing the word-error-rate of ASR systems.*
+> https://jitsi.github.io/jiwer/cli/
+
+There are plenty implementation of WER (Word Error Rate) and CER (Character Error Rate) calculation in Python but not many in NodeJs.
```
Can you detail how we build and use this package?

lutangar replied:
Not sure what the best course of action for the build instructions, should I  just mention the `build` command ? 
As for the Python requirements I'll add `"preinstall": "pip install -r requirements.txt"` to the scripts.

Chocobozzz replied:
> should I just mention the build command ?

Yep

Chocobozzz commented on packages/tests/src/core-utils/date.ts:
```diff
@@ -0,0 +1,29 @@
+import { millisecondsToTime, secondsToTime } from '@peertube/peertube-core-utils'
```
Can you also add a barrel file `index.ts` in `core-utils` directory?

lutangar replied:
Alright, but what's this barrel file should export ?

Chocobozzz replied:
`export * from './date.js'`

lutangar replied:
Humm... I'm not sure to understand the purpose of these in the tests folder. In some of the tests folder (for example in `external-plugins`) I'm seeing  `imports` instead of `exports` in the barrel file which might make more sense :thinking:

Chocobozzz replied:
Oup's sorry my bad, you're right!

lutangar replied:
Should I go with a barrel file with `imports` or with no barrel file ?

Chocobozzz replied:
Use `imports` for consistency, even if I plan to remove them in the future (I try to progressively remove barrel files from the project)

lutangar replied:
Well If it's a progressive work, we could begin with not adding this barrel file if it's not necessary !

Chocobozzz commented on packages/tests/src/jiwer/jiwer-cli.spec.ts:
```diff
@@ -0,0 +1,48 @@
+/* eslint-disable max-len */
```
Please remove `spec` from filename for consistency

Chocobozzz commented on packages/tests/src/transcription/subtitle.spec.ts:
```diff
@@ -0,0 +1,33 @@
+import { expect } from 'chai'
+
+describe('srt to txt', function () {
+  it(`Should `, function () {
```
Miss the should content :)

Chocobozzz commented on packages/tests/src/transcription/transcript/transcript-file.spec.ts:
```diff
@@ -1,26 +1,44 @@
 import { buildAbsoluteFixturePath } from '@peertube/peertube-node-utils'
 
 describe('Transcript File', function () {
+  const transcriptFileDirectory = join(tmpdir(), 'peertube-transcription/transcript-file/')
```
```suggestion
  const transcriptFileDirectory = join(tmpdir(), 'peertube-transcription', 'transcript-file')
```

lutangar replied:
Thanks :+1:

Chocobozzz commented on packages/tests/src/transcription/whisper/transcriber/openai-transcriber.spec.ts:
```diff
@@ -1,25 +1,38 @@
-  const shortVideoPath = buildAbsoluteFixturePath('video_short.mp4')
-  const frVideoPath = buildAbsoluteFixturePath('transcription/videos/communiquer-lors-dune-classe-transplantee.mp4')
-
+  const transcriptDirectory = join(tmpdir(), 'peertube-transcription/transcriber/')
```
Same suggestion as the previous comment for the `join` :)

Chocobozzz commented on packages/transcription/src/benchmark.ts:
```diff
@@ -0,0 +1,139 @@
+})
+
+void (async () => {
+  const logger = createLogger()
```
If you don't need colors or specific formatter I think you can directly use `console.log`

Chocobozzz commented on packages/transcription/src/transcription-model.ts:
```diff
@@ -16,4 +19,16 @@ export abstract class TranscriptionModel {
+  }
+
+  static fromPath (path: string) {
+    assert(existsSync(path), `${path} doesn't exist.`)
```
Please avoid sync calls

Chocobozzz commented on packages/transcription/src/whisper/transcriber/ctranslate2-transcriber.ts:
```diff
@@ -1,47 +1,49 @@
-    const { baseName } = getFileInfo(mediaFilePath)
 
     if (model.path) {
       await lstat(model.path).then(stats => stats.isDirectory())
```
Not sure to understand what's the point of this line?

lutangar replied:
You're right, let's turn this into an assertion.

lutangar committed:
fix(transcription): use ms in segment timings

lutangar committed:
chore(transcription): simplify most of the tests

lutangar committed:
chore(transcription): remove slashes when building path with join

lutangar committed:
chore(transcription): assert path to model is a directory for CTranslate2 transcriber

chore(transcription): ctranslate2 assertion

chore(transcription): ctranslate2 assertion

lutangar committed:
chore(transcription): make fromPath method async

lutangar committed:
chore(transcription): add preinstall script for Python dependencies

lutangar committed:
chore(transcription): add download and unzip utils functions

chore(transcription): add download and unzip utils functions

chore(transcription): download & unzip models fixtures
